### PR TITLE
Update cpplint version

### LINKED ISSUE #309: cpplint incorrectly labels C++17 headers as C headers
Consider this example:

```c++
// Copyright 2021 Jacob Perron
#include <iostream>
#include <optional>

int main(void)
{
  return 0;
}
```

IIUC, following the [Google C++ guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes), it has the correct, alphabetically sorted, list of included C++ headers.

However, ament_cpplint complains that `<optional>` is a C header and should appear *before* `<iostream>`.

I've noticed that the same issue happens for all C++17 headers, e.g. `<filesystem>`. cpplint doesn't seem to be aware of them. Perhaps there is a patch upstream to fix this, but I haven't taken a look.

### PR DESCRIPTION
**Pull request by jacobperron**
This PR pulls in a different [fork of cpplint](https://github.com/cpplint/cpplint) that seems to be better maintained. It also contains patches to fix #309.

I've copied the upstream file (8f925b0) and then applied relevant historical patches from our fork (1754c5f). I looked at applying the following commits from our fork, but they seem to no longer apply:

- https://github.com/ament/ament_lint/pull/67
- https://github.com/ament/ament_lint/pull/144
- https://github.com/ament/ament_lint/pull/249

---

Full CI for all connected PRs:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15926/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10601/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16250/)

PRs:
- https://github.com/ros/class_loader/pull/192
- https://github.com/ros2/demos/pull/551
- https://github.com/ros2/performance_test_fixture/pull/16
- https://github.com/ros2/rcl_logging/pull/84
- https://github.com/ros2/rclcpp/pull/1859
- https://github.com/ros2/rclpy/pull/877
- https://github.com/ros2/rcpputils/pull/158
- https://github.com/ros2/rcutils/pull/349
- https://github.com/ros/resource_retriever/pull/69
- https://github.com/ros2/rmw_connextdds/pull/69
- https://github.com/ros2/rmw_fastrtps/pull/574
- https://github.com/ros2/rmw_cyclonedds/pull/363
- https://github.com/ros2/rmw_dds_common/pull/55
- https://github.com/ros2/rmw_implementation/pull/203
- https://github.com/ros/robot_state_publisher/pull/186
- https://github.com/ros2/rosidl_typesupport_fastrtps/pull/84
- https://github.com/ros-perception/image_common/pull/221
- https://github.com/ros2/realtime_support/pull/113
- https://github.com/ros2/rviz/pull/818
- https://github.com/ros2/system_tests/pull/493
- https://github.com/ros2/urdf/pull/30
- https://github.com/ros2/geometry2/pull/497
- https://github.com/ros2/rosidl/pull/644
- https://github.com/ros2/rosbag2/pull/947

### CODE REVIEW

**Comment by jacobperron** [comment]
Testing linters only:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15247/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9966/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12929/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15440/)

**Comment by homalozoa** [comment]
Any update?

File: ament_cpplint/ament_cpplint/cpplint.py [hunk 1]
```diff
@@ -4729,28 +5322,12 @@ def CheckLanguage(filename, clean_lines, linenum, file_extension,
...
-    whitelist = [
-      'std::chrono_literals',
-      'std::complex_literals',
-      'std::literals',
-      'std::literals::chrono_literals',
-      'std::literals::complex_literals',
-      'std::literals::string_literals',
-      'std::placeholders',
-      'std::string_literals',
-    ]
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply ament_cpplint/ament_cpplint/cpplint.py]: Good question. I think we should keep the whitelist. I've reapplied the patches to and included https://github.com/ament/ament_lint/pull/67 this time.

**Comment by jacobperron** [comment]
So, with this update there's 2000+ cpplint failures across ros2 repositories related to include order. It appears the linter has become more strict about order. Most of the failures come from generated headers (from rosidl), but regardless I think we'll have to touch a large number of files to appease the linter. I don't mind going through and updating headers as necessary (@ament/team thoughts?) Here's an example of what most diffs may look like:

```diff
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <future>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <utility>
+
 #include <rclcpp/exceptions.hpp>
 #include <rclcpp/macros.hpp>
 #include <rclcpp/node_interfaces/node_base_interface.hpp>
 #include <rclcpp/node_interfaces/node_logging_interface.hpp>
 #include <rclcpp/node_interfaces/node_graph_interface.hpp>
 #include <rclcpp/logger.hpp>
 #include <rclcpp/time.hpp> #include <rclcpp/time.hpp>
 #include <rclcpp/waitable.hpp>
 
-#include <rosidl_runtime_c/action_type_support_struct.h>
+#include "rosidl_runtime_c/action_type_support_struct.h"
 #include <rosidl_typesupport_cpp/action_type_support.hpp>
 
-#include <algorithm>
-#include <chrono>
-#include <functional>
-#include <future>
-#include <map>
-#include <memory>
-#include <mutex>
-#include <string>
-#include <utility>
-
```

Note, that custom C-headers in angle brackets are considered system headers, so we should use double-quotes instead.

**Comment by clalancette** [comment]
My opinion:

If you are willing to do the work and have time to do it, I'm fine with the upgrade.  It's just going to require a large number of PRs to be merged at once :).

**Comment by jacobperron** [comment]
I think I've updated all the core repos.  I've updated this tickets description with a list of PRs and CI.

**Pull request merged**